machine_translation/__init__.py

```diff
@@ -45,131 +46,231 @@ def main(config, tr_stream, dev_stream, use_bokeh=False):
+            extensions.append(
+                Plot('Cs-En', channels=[['decoder_cost_cost']],
+                     after_batch=True))
+
+        # Set up training algorithm
+        logger.info("Initializing training algorithm")
+        algorithm = GradientDescent(
+            cost=cost, parameters=cg.parameters,
+            step_rule=CompositeRule([StepClipping(config['step_clipping']),
+                                     eval(config['step_rule'])()])
+        )
+
+        # Initialize main loop
+        logger.info("Initializing main loop")
+        main_loop = MainLoop(
+            model=training_model,
+            algorithm=algorithm,
+            data_stream=tr_stream,
+            extensions=extensions
+        )
+
+        # Train!
+        main_loop.run()
+
+    elif mode == 'translate':
+
+        # Create Theano variables
+        logger.info('Creating theano variables')
+        sampling_input = tensor.lmatrix('source')
+
+        # Get test set stream
+        test_stream = get_dev_stream(
+            config['test_set'], config['src_vocab'],
+            config['src_vocab_size'], config['unk_id'])
+        ftrans = open(config['test_set'] + '.trans.out', 'w')
+
+        # Helper utilities
+        sutils = SamplingBase()
+        unk_idx = config['unk_id']
+        src_eos_idx = config['src_vocab_size'] - 1
+        trg_eos_idx = config['trg_vocab_size'] - 1
+
+        # Get beam search
         logger.info("Building sampling model")
         sampling_representation = encoder.apply(
             sampling_input, tensor.ones(sampling_input.shape))
         generated = decoder.generate(sampling_input, sampling_representation)
-        search_model = Model(generated)
         _, samples = VariableFilter(
             bricks=[decoder.sequence_generator], name="outputs")(
                 ComputationGraph(generated[1]))  # generated[1] is next_outputs
+        beam_search = BeamSearch(samples=samples)
+
+        logger.info("Loading the model..")
+        model = Model(generated)
+        loader = LoadNMT(config['saveto'])
+        loader.set_model_parameters(model, loader.load_parameters())
+
+        # Get target vocabulary
+        trg_vocab = _ensure_special_tokens(
+            pickle.load(open(config['trg_vocab'])), bos_idx=0,
+            eos_idx=trg_eos_idx, unk_idx=unk_idx)
+        trg_ivocab = {v: k for k, v in trg_vocab.items()}
+
+        logger.info("Started translation: ")
+        total_cost = 0.0
+
+        for i, line in enumerate(test_stream.get_epoch_iterator()):
+
+            seq = sutils._oov_to_unk(
+                line[0], config['src_vocab_size'], unk_idx)
+            input_ = numpy.tile(seq, (config['beam_size'], 1))
+
+            # draw sample, checking to ensure we don't get an empty string back
+            trans, costs = \
+                beam_search.search(
+                    input_values={sampling_input: input_},
+                    max_length=3*len(seq), eol_symbol=src_eos_idx,
```

Comment:
@orhanf I'm not sure about this but when I tested on a small dataset, I saw the ```</S``` symbol everywhere in the middle of translated sentences. After changing the ```src_eos_idx``` to 0, the problem was fixed. I'm gonna double check my result again then.